File: python/hand_of_straights.py
import heapq


class Solution:
    def isNStraightHand(self, hand: list[int], groupSize: int) -> bool:
        if len(hand) % groupSize != 0:
            return False

        count: dict[int, int] = dict()
        for i in hand:
            if i in count:
                count[i] += 1
            else:
                count[i] = 1

        minHeap: list[int] = list(count.keys())
        heapq.heapify(minHeap)

        while len(minHeap) > 0:
            first: int = minHeap[0]
            for n in range(first, first + groupSize):
                if n not in count:
                    return False
                count[n] -= 1
                if count[n] == 0:
                    if n != minHeap[0]:
                        return False
                    heapq.heappop(minHeap)

        return True


def test_isNStraightHand_case_1():
    # arrange
    hand: list[int] = [1, 2, 3, 6, 2, 3, 4, 7, 8]
    groupSize: int = 3
    expected: bool = True

    # act
    actual = Solution().isNStraightHand(hand, groupSize)

    # assert
    assert actual == expected


def test_isNStraightHand_case_2():
    # arrange
    hand: list[int] = [1, 2, 3, 4, 5]
    groupSize: int = 4
    expected: bool = False

    # act
    actual = Solution().isNStraightHand(hand, groupSize)

    # assert
    assert actual == expected

File: rust/src/code/hand_of_straights.rs
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap};

struct Solution {}

impl Solution {
    // NOTE: time complexity O(nlogn)
    pub fn is_n_straight_hand(hand: Vec<i32>, group_size: i32) -> bool {
        if hand.len() % group_size as usize != 0 {
            return false;
        }

        let mut count: HashMap<i32, i32> = HashMap::new();
        for i in hand {
            count.entry(i).and_modify(|x| *x += 1).or_insert(1);
        }

        let mut min_heap: BinaryHeap<Reverse<i32>> = BinaryHeap::new();
        for k in count.keys() {
            min_heap.push(Reverse(*k));
        }

        while let Some(Reverse(first)) = min_heap.peek() {
            for number in *first..(*first + group_size) {
                if let Some(v) = count.get_mut(&number) {
                    *v -= 1;
                    if *v == 0 {
                        if let Some(Reverse(f)) = min_heap.peek() {
                            if number != *f {
                                return false;
                            }
                            min_heap.pop();
                        }
                    }
                } else {
                    return false;
                }
            }
        }

        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_n_straight_hand_case_1() {
        // arrange
        let hand = vec![1, 2, 3, 6, 2, 3, 4, 7, 8];
        let group_size = 3;
        let expected = true;

        // act
        let actual = Solution::is_n_straight_hand(hand, group_size);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_is_n_straight_hand_case_2() {
        // arrange
        let hand = vec![1, 2, 3, 4, 5];
        let group_size = 4;
        let expected = false;

        // act
        let actual = Solution::is_n_straight_hand(hand, group_size);

        // assert
        assert_eq!(actual, expected);
    }
}
